### PR TITLE
Add modtools/fire-rate

### PR DESCRIPTION
**Pull request by Atkana**
The modtool allows for a modder to alter the fire rate of ranged weapons based on different criteria.

I'm aware that several people have created similar tools, but nobody's submitted anything to be included in DFhack. Since I've just published a mod including one I made for it, I thought I might as well offer it as a contender.
Compared to some of the other modtools it does feel like a bit of a misfit (like for example, most don't replace the game mechanics with its own), but that down to what's required to make it work.
One thing the script is lacking is a mode that replicates vanilla-style fire rate rolls. I attempted to do some science to figure out how it works, but I couldn't figure out how to replicate the results with my limited maths abilities. If the lack of this style of rolls is a major sticking point, it would have to be up to somebody else to add it in.

(Hopefully this PR works fine despite my outdated fork - I'm limited to using the github site for management, which requires deleting the entire fork to update without messing up credits, and I'm currently in PR limbo with a different PR, so can't delete the fork to update it)

### CODE REVIEW

File: modtools/fire-rate.lua
Comment (myk002) [inline modtools/fire-rate.lua:6]: Typo: think -> thing

File: modtools/fire-rate.lua
Comment (myk002) [inline modtools/fire-rate.lua:184]: This if block needs a return statement, otherwise the (invalid) value will be used a few lines down

**Review by myk002 — COMMENTED** [review]
When you get the chance, please update changelog.txt with this new script. Also, could you wrap the text at about the 80th column? Since of the lines are really long.

File: modtools/fire-rate.lua
Comment (Atkana) [reply modtools/fire-rate.lua]: fixed

File: modtools/fire-rate.lua
Comment (Atkana) [reply modtools/fire-rate.lua:190]: oops - fixed

**Comment by Atkana** [comment]
> When you get the chance, please update changelog.txt with this new script. Also, could you wrap the text at about the 80th column? Since of the lines are really long.

Unfortunately as I mentioned in the PR, my fork is fairly outdated (from back at ~`0.47.04-r3` from the looks of things), so I don't know if an edit to the changelog would successfully merge. I can at least write something out for somebody else to add if/when this gets added:
```
## New Scripts
- `modtools/fire-rate`: allows users to adjust the rate of fire for ranged attacks
```

I made some edits to shorten some of the lines somewhat, but how necessary is that? In some cases it has hurt readability, and as far as I understand it's only the github site itself that lacks wrapping in its view mode (edit mode wraps, raw mode wraps*, the docs wrap, the DFhack console wraps).
(*I think that ones technically down to the browser, but shush :p)

**Comment by myk002** [comment]
> my fork is fairly outdated

On your fork page (https://github.com/Atkana/scripts/tree/add-fire-rate) you should see a little link for "fetch upstream". That should get you up to date.

> I made some edits to shorten some of the lines somewhat, but how necessary is that?

It's for the sanity of those who have to review and maintain the code. Long lines make code diffs much harder to read, and even if the code wraps, indentation of the wrapped lines won't necessarily match the logical containing code block. The usual standard for code is 80 columns (with some exceptions, like for golang, but we don't use golang).

**Comment by Atkana** [comment]
I'm not sure if it's changed since last time I used it, but as far as I'm aware the Fetch upstream option on the site makes a commit to add all the changes to the fork, meaning it'll mess up the contributions for the 607 commits in need of merging.
I think it happened last time I tried, but there were far fewer commits to be messed up. The thought of doing that now I know that happens and on a larger scale stresses me out way too much to try. I'm not sure how far the permission extends with the "Allow edits by maintainers" option enabled, but if it's in the means of somebody else to do it, they're welcome to.

**Comment by myk002** [comment]
merging only has issues when there are conflicting changes, of which this PR has none. It's not a huge problem, though. I can add the changelog entry after merge. I wrapped the code for you. What is a good way for me to test the script in a running game?

**Comment by Atkana** [comment]
> What is a good way for me to test the script in a running game?

It's rigged up so that both modes can be tested in the arena - fort mode rates are used in the general arena, and adventure mode rates are used while assuming control of a character. Here are a few tests to demonstrate different things (make sure to run the commands when the world is loaded):

Changes between modes. In this test bows will fire every tick in fortress mode, but take forever to recover from in adventure mode. For the adventure mode test it's best if there's also another creature active so there will be reports about its actions during your recovery time.
`modtools/fire-rate -item WEAPON:ITEM_WEAPON_BOW -fortBase 1 -advBase 10000`

Effects of skill. In this, an low skilled crossbowmen will fire really slowly in fort mode, but at level 10 they'll fire at 1 shot per tick.
`modtools/fire-rate -item WEAPON:ITEM_WEAPON_CROSSBOW -fortBase 1500 -fortSkillFactor 150 -fortCap 1`

Effects of material. Crossbows made of gold will always fire at the fastest fire rate possible. If the previous test is still active you can compare the rates of an unskilled crossbow user using a regular one vs a golden one.
`modtools/fire-rate -item WEAPON:ITEM_WEAPON_CROSSBOW -material INORGANIC:GOLD -fortBase 1 -advBase 1`

Materials + interactions. This'll make a dwarf spitting in adventure mode take a really long time to recover from (again, with this sort of test it's best for another creature to also be active):
`modtools/fire-rate -throw -material CREATURE:DWARF:SPIT -advBase 10000`

... I'm just now realising (after writing up a test for throwing weapons that wouldn't work) that I never thought to allow for adjustments when specifically throwing *items*. I was so wrapped up with the concept of people wanting to buff firing rates that I didn't consider they might want to also nerf throwing rates. I'll have to pencil that in as something to consider adding in a future version.

File: modtools/fire-rate.lua
Comment (myk002) [inline modtools/fire-rate.lua:141]: in lua, the preferred form would be:
`if not material_token then`
that is, just test the var for "truthiness" and don't use parentheses unless you have to for grouping

File: modtools/fire-rate.lua
Comment (myk002) [inline modtools/fire-rate.lua:145]: `if not data`

File: modtools/fire-rate.lua
Comment (myk002) [inline modtools/fire-rate.lua:157]: can remove parens

File: modtools/fire-rate.lua
Comment (myk002) [inline modtools/fire-rate.lua:175]: This function could be considerably shorter:
```
return weapon_data[weapon_token..separator..(material_token or '')] or
        weapon_data[weapon_token] or
        register_weapon(weapon_token)
```
with all comments becoming comments about the function as a whole

File: modtools/fire-rate.lua
Comment (myk002) [inline modtools/fire-rate.lua:324]: consider: `local subtype_token = item_token:split(':')[2]`

File: modtools/fire-rate.lua
Comment (myk002) [inline modtools/fire-rate.lua:298]: It might be clearer to put the magic number 9 in a named constant so it's more obvious what is going on here

**Review by myk002 — CHANGES_REQUESTED** [review]

File: modtools/fire-rate.lua
Comment (Atkana) [reply modtools/fire-rate.lua:10]: That's a good idea - I've added it

File: modtools/fire-rate.lua
Comment (Atkana) [reply modtools/fire-rate.lua:46]: That makes more sense. Edited.

File: modtools/fire-rate.lua
Comment (Atkana) [reply modtools/fire-rate.lua:187]: I will usually go out of my way to make things longer than necessary if it means making the code easier to follow, though since most of the clarity comes from the comments in this case, I'm fine making this change (and have done so).

File: modtools/fire-rate.lua
Comment (Atkana) [reply modtools/fire-rate.lua]: That's a good idea. Made that edit.

File: modtools/fire-rate.lua
Comment (Atkana) [reply modtools/fire-rate.lua:318]: I was originally going to object to that change on the grounds that it unnecessarily obfuscates where the split function was coming from, but then I realised it also doesn't come from `utils` either. Where *does* split come from? And what's the point of `utils` having the `split_string` function?

File: modtools/fire-rate.lua
Comment (Atkana) [reply modtools/fire-rate.lua:2]: I've been burned more often than not by things having private functions that I need to then replicate myself, so I tend to leave everything available to use and let the other scripters discern what to use.

File: modtools/fire-rate.lua
Comment (Atkana) [reply modtools/fire-rate.lua:146]: Addressing a few of the formatting points: I tend to try and make things as clear as possible (even if it means slightly longer code), because I know that one of the main ways for beginners getting in to making DFhack scripts is to read existing ones to learn from.

File: modtools/fire-rate.lua
Comment (myk002) [reply modtools/fire-rate.lua:318]: sorry, I thought I replied to this earlier. the `string:split()` function is defined in `dfhack.lua` so it's available to all scripts:
https://github.com/DFHack/dfhack/blob/develop/library/lua/dfhack.lua#L413

the function used to be defined in `utils.lua`. The definition was recently (within the last year) moved to `dfhack.lua` so it was accessible as a `string` function along with the other related string functions that `dfhack.lua` provides.

**Review by myk002 — APPROVED** [review]

**Comment by myk002** [comment]
Note to merger -- make a changelog entry after merge

**Pull request merged**
